File: src/data/courses/dsa/graphs/topological-sort/topological-sort-dfs/code/java/TopologicalSortDfs.java
// ============================================================
// Topological Sort — DFS
// ============================================================
import java.util.*;
class Solution {
    public List<Integer> topologicalSort(int n, List<List<Integer>> adj) {
        boolean[] visited = new boolean[n];
        Deque<Integer> stack = new ArrayDeque<>();
        for (int i = 0; i < n; i++)
            if (!visited[i]) dfs(i, adj, visited, stack);
        return new ArrayList<>(stack);
    }
    void dfs(int node, List<List<Integer>> adj, boolean[] visited, Deque<Integer> stack) {
        visited[node] = true;
        for (int nb : adj.get(node)) if (!visited[nb]) dfs(nb, adj, visited, stack);
        stack.push(node);
    }
}

File: src/data/courses/dsa/graphs/topological-sort/topological-sort-dfs/code/python/topological_sort_dfs.py
# ============================================================
# Topological Sort — DFS
# ============================================================
class Solution:
    def topologicalSort(self, n, adj):
        visited = [False] * n
        stack = []
        def dfs(node):
            visited[node] = True
            for nb in adj[node]:
                if not visited[nb]: dfs(nb)
            stack.append(node)  # Postorder
        for i in range(n):
            if not visited[i]: dfs(i)
        return stack[::-1]  # Reverse postorder = topo order
